lithium/lithium.py

```diff
@@ -1,1215 +1,1296 @@
+    name = "replace-properties-by-globals"
+
+    # This strategy attempts to remove members, such that other strategies can
+    # then move the lines outside the functions.  The goal is to rename
+    # variables at the same time, such that the program remains valid, while
+    # removing the dependency on the object on which the member is part of.
+    #
+    #   function Foo() {
+    #     this.list = [];
+    #   }
+    #   Foo.prototype.push = function(a) {
+    #     this.list.push(a);
+    #   }
+    #   Foo.prototype.last = function() {
+    #     return this.list.pop();
+    #   }
+    #
+    # Which might transform the previous example to something like:
+    #
+    #   function Foo() {
+    #     list = [];
+    #   }
+    #   push = function(a) {
+    #     list.push(a);
+    #   }
+    #   last = function() {
+    #     return list.pop();
+    #   }
+    #
+    def run(self, testcase, interesting, tempFilename):
+        chunkSize = min(self.minimizeMax, 2 * largestPowerOfTwoSmallerThan(len(testcase.parts)))
+        finalChunkSize = max(self.minimizeMin, 1)
+
+        origNumChars = 0
+        for line in testcase.parts:
+            origNumChars += len(line)
+
+        numChars = origNumChars
+        while 1:
+            numRemovedChars, testcase = self.tryMakingGlobals(chunkSize, numChars, testcase, interesting, tempFilename)
+            numChars -= numRemovedChars
 
-    writeTestcase(testcaseFilename)
+            last = (chunkSize <= finalChunkSize)
 
-    print "=== LITHIUM SUMMARY ==="
-    if finalChunkSize == 1 and minimizeRepeat != "never":
-        print "  Removing any single " + atom + " from the final file makes it uninteresting!"
+            if numRemovedChars and (self.minimizeRepeat == "always" or (self.minimizeRepeat == "last" and last)):
+                # Repeat with the same chunk size
+                pass
+            elif last:
+                # Done
+                break
+            else:
+                # Continue with the next smaller chunk size
+                chunkSize >>= 1
 
-    print "  Initial size: " + quantity(origNumChars, "character")
-    print "  Final size: " + quantity(numChars, "character")
-    print "  Tests performed: " + str(testCount)
-    print "  Test total: " + quantity(testTotal, atom)
+        log.info("  Initial size: %s", quantity(origNumChars, "character"))
+        log.info("  Final size: %s", quantity(numChars, "character"))
 
+        return 0, (finalChunkSize == 1 and self.minimizeRepeat != "never"), testcase
 
-def tryMakingGlobals(chunkSize, numChars):
-    """Make a single run through the testcase, trying to remove chunks of size chunkSize.
 
-    Returns True iff any chunks were removed."""
+    def tryMakingGlobals(self, chunkSize, numChars, testcase, interesting, tempFilename):
+        """Make a single run through the testcase, trying to remove chunks of size chunkSize.
 
-    global parts
+        Returns True iff any chunks were rem
```

Comment:
I'm assuming you mean L946... I put it here because we're printing testcase data (which may or may not be UTF-8 data, but it's informational only). I did find one more on L798! I think that's all, I've gone through every `log.info` and `description = ...` now. Thanks!

--cutBefore and --cutAfter also decode UTF-8 to print the defaults for the help message (L1175-L1179), but those strings are constants so it makes sense to error if they're invalid.